### ISSUE TITLE
Support for `@PreAuthorize` spring-security annotation

### ISSUE THREAD

**Issue by gmarziou** [springfox/springfox#1655]
Hello,

I'm using Springfox 2.6.1 in JHipster apps and this is a feature request.

It would be nice if Springfox could support spring-security annotations, especially  `@PreAuthorize` .
It could be:
- Exposing in the swagger UI the requirements to access a resource of its methods
- Expose in the swagger spec only the resources and methods allowed for current user

Thanks

**Comment by dilipkrish** [comment]
Hello @gmarziou!

If I understand you correctly, are you suggesting we try and use @PreAuthorize to feed into the securityContexts and securitySchemes in to the Docket?

**Comment by gmarziou** [comment]
Yes that's correct, it seems to me that currently  the securityContexts and securitySchemes mainly deal with authentication and I would like to be able to document also authorizations.

The other point is that the swagger spec generated dynamically does not take into account the authorizations of the current user and so is exposing the complete API even though the current user is not authorized to use it. To me that's a reason why some people are reluctant to enable swagger in production.

**Comment by fredboutin** [comment]
This would really be useful for us! Any updates on whether you guys will integrate this feature?

**Comment by TomJahncke** [comment]
I agree this would be a very useful feature and would like to see it.  You might not want this feature on by default as this could leak security details, such as what role you need to access security services.

**Comment by Hronom** [comment]
Hello any progress on this?

**Comment by mahesh-uma** [comment]
Hello any progress on this

**Comment by dilipkrish** [comment]
Would love for a contribution. Not very familiar with how spring-security works and how that maps to swagger-ui/editor

**Comment by raderio** [comment]
https://github.com/swagger-api/swagger-core/issues/2353

**Comment by jkeruzec** [comment]
About security and Docket, one solution :  
https://github.com/springfox/springfox/issues/1600

**Comment by wojciech-soltys** [comment]
Solution for antMatchers: https://blog.codecentric.de/2018/11/springfoxswaggerextension/

**Comment by kevinhooke** [comment]
Also interested in support for @PreAuthorize() annotated endpoints to appear in the generated docs

**Comment by chriswhite199** [comment]
Adapting the antMatcher link posted by @wojciech-soltys, the following works for my purposes (but by **NO MEANS** is exhaustive for all spring security annotations)

```
/**
 * Parse out Spring Security {@link PreAuthorize} annotations and add in operation notes section to convey constraints
 */
@Component
@Order(SwaggerPluginSupport.SWAGGER_PLUGIN_ORDER)
@Slf4j
public class OperationNotesResourcesReader implements OperationBuilderPlugin {
    @Override
    public void apply(final OperationContext context) {
        // Look for @PreAuthorize on method, otherwise look on controller
        Optional.ofNullable(context.findAnnotation(PreAuthorize.class)
            .or(context.findControllerAnnotation(PreAuthorize.class))
            .orNull())
            .ifPresent(preAuth -> context.operationBuilder()
                .notes("**Security @PreAuthorize expression:** `" + preAuth.value() + "`"));
    }

    @Override
    public boolean supports(final DocumentationType delimiter) {
        return SwaggerPluginSupport.pluginDoesApply(delimiter);
    }
}
```

![Screen Shot 2019-10-10 at 2 31 51 PM](https://user-images.githubusercontent.com/1356204/66608448-6d57a800-eb6b-11e9-9b72-8b84b5033de4.png)

**Comment by vlad2** [comment]
Using java's Optional, the class above becomes:

```java
package ro.vdin;

import org.springframework.context.annotation.Profile;
import org.springframework.core.annotation.Order;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.stereotype.Component;
import springfox.documentation.spi.DocumentationType;
import springfox.documentation.spi.service.OperationBuilderPlugin;
import springfox.documentation.spi.service.contexts.OperationContext;
import springfox.documentation.swagger.common.SwaggerPluginSupport;

/**
 * Parses out Spring Security {@link PreAuthorize} annotations and adds them in the notes section.
 */
@Component
@Profile({"default" /* This disables Swagger in unittests */, "!prod"})
@Order(SwaggerPluginSupport.SWAGGER_PLUGIN_ORDER)
public class SwaggerPreAuthorize implements OperationBuilderPlugin {
    @Override
    public void apply(final OperationContext context) {
        // Look for @PreAuthorize on method, otherwise look on controller
        context.findAnnotation(PreAuthorize.class)
                .or(() -> context.findControllerAnnotation(PreAuthorize.class))
                .ifPresent(preAuth -> context.operationBuilder()
                        .notes(getNotes(preAuth)));
    }

    private String getNotes(PreAuthorize preAuth) {
        String preauthExpr = preAuth.value();
        return "**Security @PreAuthorize expression:** `" + preauthExpr + "`";
    }

    @Override
    public boolean supports(final DocumentationType delimiter) {
        return SwaggerPluginSupport.pluginDoesApply(delimiter);
    }
}
```

**Comment by lpghartz** [comment]
I know this is pretty old, but just to throw my hat into the discussion, related to this.  There's two things mentioned above, and they're very different. Documenting the security requirements for an API vs controlling the visibility of the API based on the evaluated security constraints at request-time.  The former is useful in fairly limited cases -- as it requires that the end-user of the API understand the implementation details of the authorization mechanism used in the server. For a multitude of reasons, not the least of which is turning implementation details into public (and therefore largely immutable) details. 

There's obviously ways covered above for how to handle automatically documenting the @PreAuthorize, right or wrong as it may be for any given use case, but the second question in the original issue hasn't been addressed. 

From what I can tell, that second question -- filtering the specifications by permission -- is not possible, even with Springfox 3.0? I seems that the specification is generated at startup time, and is never evaluated again, rather than at request-time. So AuthZ-based filtering can't occur, nor can any multi-tenant filtering/modifications, or anything at at all where the resulting documentation is specific to the context of its request (user context, time context, etc?)

Does anyone know if this is an actual fundamental limitation in Springfox, or just a limitation in the documentation not calling out existing extension points that are utilized at runtime?

**Comment by mdbudnick** [comment]
+1 for exposing the authorization required by the endpoints. With intuitive responsibility names this can be a very useful feature for understanding which permissions are required to use an endpoint.

Will be looking into @chriswhite199 implementation
